### PR TITLE
curvefs/metaserver: added metaserver statistics for heartbeat.

### PR DESCRIPTION
**Pull request by Wine93**

### CODE REVIEW

File: curvefs/conf/metaserver.conf [hunk 1]
```diff
@@ -187,3 +187,11 @@ mds.heartbeat_timeoutMs=1000
 partition.clean.scanPeriodSec=10
 # partition clean manager delete inode every inodeDeletePeriodMs
 partition.clean.inodeDeletePeriodMs=500
+
+#
+# storage settings
+#
+# metaserver max memory bytes (default: 5GB)
+storage.max_memory_bytes=5368709120
```
Comment (cw123) [inline curvefs/conf/metaserver.conf:195]: how about memory quota

File: curvefs/src/metaserver/storage/config.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#define CURVEFS_SRC_METASERVER_STORAGE_CONFIG_H_
+
+#include <string>
+
+namespace curvefs {
+namespace metaserver {
+namespace storage {
+
+struct StorageOptions {
+    uint64_t MaxMemoryBytes;
```
Comment (cw123) [inline curvefs/src/metaserver/storage/config.h:26]: maxMemoryBytes

File: curvefs/src/metaserver/storage/config.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    uint64_t MaxDiskQuotaBytes;
```
Comment (cw123) [inline curvefs/src/metaserver/storage/config.h:27]: maxDiskQuotaBytes

File: curvefs/src/metaserver/storage/storage.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+namespace curvefs {
+namespace metaserver {
+namespace storage {
+
+using ::curvefs::metaserver::storage::KVStorage;
+
+static std::shared_ptr<KVStorage> kKVStorage;
+
+bool InitStorage(StorageOptions options) {
+    kKVStorage = std::make_shared<MemoryStorage>(options);
```
Comment (cw123) [inline curvefs/src/metaserver/storage/storage.cpp:37]: memory storage to kv storage ?

File: curvefs/src/metaserver/storage/storage.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    return true;
```
Comment (cw123) [inline curvefs/src/metaserver/storage/storage.cpp:38]: it always return true, can return void instead

File: curvefs/src/metaserver/storage/memory_storage.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+bool MemoryStorage::GetStatistics(StorageStatistics* statistics) {
+    statistics->MaxMemoryBytes = options_.MaxMemoryBytes;
+    statistics->MaxDiskQuotaBytes = options_.MaxDiskQuotaBytes;
+
+    uint64_t vmRSS;
+    if (!GetProcMemory(&vmRSS)) {
+        return false;
+    }
+    statistics->MemoryUsageBytes = vmRSS;
+    statistics->DiskUsageBytes = 0;
```
Comment (cw123) [inline curvefs/src/metaserver/storage/memory_storage.cpp:45]: raft also use disk, should we pass disk usage?

File: curvefs/src/metaserver/heartbeat.cpp [hunk 7]
```diff
@@ -236,6 +239,30 @@ bool Heartbeat::GetProcMemory(uint64_t* vmRSS) {
...
+    status->set_memorythresholdbyte(statistics.MaxMemoryBytes);
+    status->set_memoryusedbyte(statistics.MemoryUsageBytes);
+    status->set_diskthresholdbyte(statistics.MaxDiskQuotaBytes);
+    status->set_diskusedbyte(0);
+    status->set_diskcopysetminrequirebyte(0);
+    if (ncopysets == 0) {
+        status->set_memorycopysetminrequirebyte(0);
+    } else {
+        status->set_memorycopysetminrequirebyte(
+            uint64_t(statistics.MemoryUsageBytes / ncopysets));
```
Comment (cw123) [inline curvefs/src/metaserver/heartbeat.cpp:261]: Why do you calculate it this way

File: curvefs/conf/metaserver.conf (hunk 1 above)
Comment (Wine93) [reply curvefs/conf/metaserver.conf:195]: > how about memory quota

agree!

File: curvefs/src/metaserver/storage/storage.cpp (hunk 4 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/storage.cpp:37]: > memory storage to kv storage ?

yes, kv is also can be in-memory storage, like `redis`.

File: curvefs/src/metaserver/storage/storage.cpp (hunk 5 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/storage.cpp:38]: > it always return true, can return void instead

agree!

**Comment by Wine93** [comment]
recheck

File: curvefs/src/metaserver/storage/config.h (hunk 2 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/config.h:33]: > maxMemoryBytes

fixed.

File: curvefs/src/metaserver/storage/config.h (hunk 3 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/config.h:34]: > maxDiskQuotaBytes

fixed.

File: curvefs/src/metaserver/storage/memory_storage.cpp (hunk 6 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/memory_storage.cpp:45]: > raft also use disk, should we pass disk usage?

done.

File: curvefs/src/metaserver/heartbeat.cpp (hunk 7 above)
Comment (Wine93) [reply curvefs/src/metaserver/heartbeat.cpp:261]: > Why do you calculate it this way

It's a simplified strategy, we can balance metaserver storage in cluster according to its guidance.

File: curvefs/src/metaserver/storage/storage.cpp (hunk 5 above)
Comment (Wine93) [reply curvefs/src/metaserver/storage/storage.cpp:38]: > it always return true, can return void instead

done.

**Comment by Wine93** [comment]
recheck

File: curvefs/src/metaserver/storage/utils.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+namespace curvefs {
+namespace metaserver {
+namespace storage {
+
+bool GetFileSystemSpaces(const std::string& path,
+                         uint64_t* capacity,
+                         uint64_t* available);
+
+bool GetProcMemory(uint64_t* vmRSS);
+
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/utils.h:37]: The function 'GetFileSystemSpaces' and 'GetProcMemory' already exist in heartbeat.cpp, reuse or deal that one.

Comment (Wine93) [reply curvefs/src/metaserver/storage/utils.h:37]: > The function 'GetFileSystemSpaces' and 'GetProcMemory' already exist in heartbeat.cpp, reuse or deal that one.

fixed

**Review by SeanHai — APPROVED** [review]

File: curvefs/src/metaserver/heartbeat.cpp [hunk 9]
```diff
@@ -247,7 +235,7 @@ int Heartbeat::BuildRequest(HeartbeatRequest* req) {
 
     uint64_t capacityKB = 0;
     uint64_t availKB = 0;
-    ret = GetFileSystemSpaces(&capacityKB, &availKB);
+    ret = GetFileSystemSpaces(storePath_, &capacityKB, &availKB);
```
Comment (cw123) [inline curvefs/src/metaserver/heartbeat.cpp:238]: this code doesn't need, use spacestatus instead.

delete the // code below

**Review by cw123 — APPROVED** [review]

**Comment by YunhuiChen** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Comment by Wine93** [comment]
recheck

**Pull request merged**